probe-rs/src/probe/jlink/swd.rs

```diff
@@ -122,20 +123,217 @@ impl SwdStatistics {
     }
 }
 
-///! Implementation of the SWD protocol for the JLink probe.
+///! Implementation of the SWD and JTAG protocols for the JLink probe.
+
+// Constant to be written to ABORT
+const JTAG_ABORT_VALUE: u64 = 0x8;
+
+// IR values for JTAG registers
+const JTAG_ABORT_IR_VALUE: u32 = 0x8;
+const JTAG_DEBUG_PORT_IR_VALUE: u32 = 0xA;
+const JTAG_ACCESS_PORT_IR_VALUE: u32 = 0xB;
+
+const JTAG_STATUS_WAIT: u32 = 0x1;
+const JTAG_STATUS_OK: u32 = 0x2;
+
+// Build a JTAG payload
+fn build_jtag_payload_and_address(transfer: &DapTransfer) -> (u64, u32) {
+    if transfer.is_abort() {
+        (JTAG_ABORT_VALUE, JTAG_ABORT_IR_VALUE)
+    } else {
+        let address = match transfer.port {
+            PortType::DebugPort => JTAG_DEBUG_PORT_IR_VALUE,
+            PortType::AccessPort => JTAG_ACCESS_PORT_IR_VALUE,
+        };
+
+        let mut payload = 0u64;
+
+        // 32-bit value, bits 35:3
+        payload |= (transfer.value as u64) << 3;
+        // A[3:2], bits 2:1
+        payload |= (transfer.address as u64 & 0b1000) >> 1;
+        payload |= (transfer.address as u64 & 0b0100) >> 1;
+        // RnW, bit 0
+        payload |= if transfer.direction == TransferDirection::Read {
+            1
+        } else {
+            0
+        };
+
+        (payload, address)
+    }
+}
+
+fn parse_jtag_response(data: &[u8]) -> u64 {
+    let mut received = 0u64;
+    for v in data.iter() {
+        received >>= 8;
+        received |= (*v as u64) << 32;
+    }
+
+    received
+}
+
+/// Perform a single JTAG tranfer and parse the results
+///
+/// Return is (value, status)
+fn perform_jtag_transfer<P: JTAGAccess + RawProtocolIo>(
+    probe: &mut P,
+    transfer: &DapTransfer,
+) -> Result<(u32, TransferStatus), DebugProbeError> {
+    // Determine what JTAG IR address and value to send
+    let (payload, address) = build_jtag_payload_and_address(transfer);
+    let data = payload.to_le_bytes();
+
+    // This is a bit confusing, but a read from any port is still
+    // a JTAG write as we have to transmit the address
+    let result = probe.write_register(address, &data[..], 35)?;
+
+    // Clock out any idle time
+    let idle_sequence = iter::repeat(false).take(transfer.idle_cycles_after as usize);
+    probe.jtag_io(idle_sequence.to_owned(), idle_sequence)?;
+
+    let received = parse_jtag_response(&result);
+
+    if transfer.is_abort() {
+        // No resposnes returned from this
+        return Ok((0, TransferStatus::Ok));
+    }
+
+    // Received value is bits [35:3]
+    let received_value = (received >> 3) as u32;
+    // Status is bits [2:0]
+    let status = (received & 0b111) as u32;
+
+    let transfer_status = match status {
+        JTAG_STATUS_WAIT => {
+            // WAIT
+            TransferStatus::Failed(DapError::WaitResponse)
+        }
+        JTAG_STATUS_OK => {
+            // OK/FAULT
```

Comment:
Is there a separate status for FAULT?

Reply:
No - and this is a bit confusing.  The value in the ARM JTAG documentation is described as OK/FAULT.

There's not a separate fault bit as unlike SWD there are no protocol level failures.  A 0/1 is always valid for every bit.  If you read an invalid register, such as one that is SWD only, in JTAG it will return OK/FAULT and an undefined value.  You're basically supposed to ensure your JTAG code doesn't make bad requests, which makes debugging a bit hard but there's not much more we can do here.

The only fault that can actually happen assuming you're reading valid registers is an overrun, which returns WAIT.